clients/src/test/java/org/apache/kafka/clients/NetworkClientTest.java

```diff
@@ -551,4 +566,61 @@ public void onComplete(ClientResponse response) {
             this.response = response;
         }
     }
+
+    // ManualMetadataUpdater with ability to keep track of failures
+    private static class TestMetadataUpdater implements MetadataUpdater {
+        private final ManualMetadataUpdater updater;
```

Comment:
Did you consider extending `ManualMetadataUpdater`? That would let us skip most of the overrides below.